server/src/main/java/org/opensearch/search/aggregations/AggregationCollectorManager.java

```diff
@@ -63,18 +63,17 @@ public ReduceableSearchResult reduce(Collection<Collector> collectors) throws IO
         }
 
         final InternalAggregations internalAggregations = InternalAggregations.from(internals);
-        // Reduce the aggregations across slices before sending to the coordinator. We will perform shard level reduce iff multiple slices
-        // were created to execute this request and it used concurrent segment search path
+        // Reduce the aggregations across slices before sending to the coordinator. We will perform shard level reduce as long as any slices
+        // were created so that we can apply shard level bucket count thresholds in the reduce phase.
         // TODO: Add the check for flag that the request was executed using concurrent search
-        if (collectors.size() > 1) {
+        if (!collectors.isEmpty() && context.isConcurrentSegmentSearchEnabled()) {
```

Comment:
`context.isConcurrentSegmentSearchEnabled()` - this check is not needed, the reduce will always be called in context of concurrent search (otherwise `collectors` are always empty. Please use the convention that is being followed:

```
if (collectors.isEmpty() == false) {
 ....
}
```

Reply:
I think what @sohami was calling out [here](https://github.com/opensearch-project/OpenSearch/pull/9085#discussion_r1287410895) is that this path is still called in the non-concurrent case since it's handled by `DefaultAggregationProcessor` now. I double checked with the debugger that this is the case:

```
context.isConcurrentSegmentSearchEnabled() = false
collectors = {ImmutableCollections$List12@16496}  size = 1
 0 = {FilterAggregator@14209} "inclass"
```

Reply:
Missed that, thank you, still why we need this check at all if we do call `context.isConcurrentSegmentSearchEnabled()` everywhere down the stack? Does it break something? My concern is that AggregationCollectorManager should be agnostic to concurrent / non-concurrent search.

Reply:
That's a fair point. I think in the context of this PR it would be good to avoid changing both the concurrent and non-concurrent paths at the same time since today `reduce` is only called for the concurrent case.
I think as a separate issue we could look into if it's safe to call `reduce` in both cases, and if so move all of the threshold checks out of `buildAggregations`.

Reply:
> I think in the context of this PR it would be good to avoid changing both the concurrent and non-concurrent paths at the same time since today `reduce` is only called for the concurrent case.

I think from the previous comment, the `reduce` is called for both BUT for non-concurrent case - at most with 1 collector so the `InternalAggregations::reduce` never called. Let dig into this particular flow, I would certainly prefer to keep `AggregationCollectorManager` clean

Reply:
`reduce` is not called for both (concurrent and non-concurrent case). `collector > 1` check was used to differentiate 2 things a) For non-concurrent case it will always be 1 and b) for concurrent case we were avoiding it being called with slice as 1. We should not call `reduce` for non-concurrent case. We can create a method in `AggregationCollectorManager` something like `AggregationReduceableSearchResult buildAggregationResult(InternalAggregations internalAggregations)`. This method can be called from `reduce` in `AggregationCollectorManager`. Then the collection manager for non concurrent and concurrent flow can override this and implement its custom logic ?

Reply:
> This method can be called from `reduce` in `AggregationCollectorManager`. Then the collection manager for non concurrent and concurrent flow can override this and implement its custom logic ?

:+1: , I think this is good idea since hierarchy of the collector managers is separated, thanks @sohami

Reply:
Took a crack at implementing this in the latest revision, let me know what you think!